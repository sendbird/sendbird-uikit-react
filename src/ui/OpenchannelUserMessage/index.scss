@import '../../styles/variables';

.sendbird-openchannel-user-message {
  position: relative;
  display: flex;
  flex-direction: row;
  width: 100%;
  &:hover {
    @include themed() {
      background-color: t(bg--1-4);
    }
    .sendbird-openchannel-user-message__context-menu {
      display: block;
    }
  }

  .sendbird-openchannel-user-message__left {
    position: relative;
    display: flex;
    width: 52px;
    min-width: 52px;

    .sendbird-openchannel-user-message__left__avatar {
      position: relative;
      display: flex;
      margin-top: 8px;
      margin-left: 24px;
    }
  }

  .sendbird-openchannel-user-message__right {
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 8px 24px 8px 12px;

    .sendbird-openchannel-user-message__right__top {
      position: relative;
      display: flex;
      margin-bottom: 2px;

      .sendbird-openchannel-user-message__right__top__sender-name {
        position: relative;
        display: flex;
      }
      .sendbird-openchannel-user-message__right__top__sent-at {
        position: relative;
        display: flex;
        margin-left: 4px;
      }
    }

    .sendbird-openchannel-user-message__right__bottom {
      position: relative;
      display: flex;
      width: 100%;

      .sendbird-openchannel-user-message__right__bottom__message {
        position: relative;
        word-break: break-all;
        white-space: pre-line;
      }
    }

    .sendbird-openchannel-user-message__right__tail {
      position: relative;
      top: 2px;
    }
  }

  .sendbird-openchannel-user-message__context-menu {
    position: absolute;
    right: 16px;
    display: none;

    .sendbird-openchannel-user-message__context-menu--icon {
      @include themed() {
        background-color: t(bg--2-5);
      }
      &:hover {
        @include themed() {
          background-color: t(primary--1-bg--7);
        }
      }
    }
  }
}

// .sendbird-openchannel-user-message__context-menu__copy {

// }

// .sendbird-openchannel-user-message__context-menu__edit {

// }

// .sendbird-openchannel-user-message__context-menu__resend {

// }

// .sendbird-openchannel-user-message__context-menu__delete {

// }
