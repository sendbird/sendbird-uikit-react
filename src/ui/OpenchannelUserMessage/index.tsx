import React, {
  useContext,
  useRef,
  useState,
  ReactElement,
  useEffect,
} from 'react';
import { UserMessage } from '@sendbird/chat/message';
import format from 'date-fns/format';
import './index.scss';

import Avatar from '../Avatar';
import ContextMenu, { MenuItems, MenuItem } from '../ContextMenu';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import Label, { LabelTypography, LabelColors } from '../Label';
import Loader from '../Loader';
import UserProfile from '../UserProfile';
import { UserProfileContext } from '../../lib/UserProfileContext';

import { useLocalization } from '../../lib/LocalizationContext';
import { copyToClipboard } from './utils';
import uuidv4 from '../../utils/uuid';
import {
  checkIsPending,
  checkIsFailed,
  isFineCopy,
  isFineEdit,
  isFineResend,
  isFineDelete,
  showMenuTrigger,
} from '../../utils/openChannelUtils';
import { getSenderFromMessage } from '../../utils/openChannelUtils';
import { isEditedMessage } from '../../utils';

interface Props {
  className?: string | Array<string>;
  message: UserMessage;
  isOperator?: boolean;
  userId: string;
  disabled?: boolean;
  showEdit(bool: boolean): void;
  showRemove(bool: boolean): void;
  resendMessage(message: UserMessage): void;
  chainTop?: boolean;
  chainBottom?: boolean;
}

export default function OpenchannelUserMessage({
  className,
  message,
  isOperator,
  userId,
  resendMessage,
  disabled,
  showEdit,
  showRemove,
  chainTop,
}: Props): ReactElement {
  if (!message || message.messageType !== 'user') {
    return null;
  }

  // hooks
  const { stringSet, dateLocale } = useLocalization();
  const { disableUserProfile, renderUserProfile } = useContext<UserProfileContext>(UserProfileContext);
  const messageRef = useRef(null);
  const avatarRef = useRef(null);
  const contextMenuRef = useRef(null);
  const [contextStyle, setContextStyle] = useState({});

  // consts
  const status = message?.sendingStatus;
  const isPending = checkIsPending(status);
  const isFailed = checkIsFailed(status);
  const sender = getSenderFromMessage(message);

  // place context menu top depending clientHeight of message component
  useEffect(() => {
    if (messageRef?.current?.clientHeight > 36) {
      setContextStyle({ top: '8px ' });
    } else {
      setContextStyle({ top: '2px' });
    }
  }, [window.innerWidth]);

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-openchannel-user-message',
      ].join(' ')}
      ref={messageRef}
    >
      <div className="sendbird-openchannel-user-message__left">
        {
          !chainTop && (
            <ContextMenu
              menuTrigger={(toggleDropdown) => (
                <Avatar
                  className="sendbird-openchannel-user-message__left__avatar"
                  src={sender.profileUrl || ''}
                  ref={avatarRef}
                  width="28px"
                  height="28px"
                  onClick={() => {
                    if (!disableUserProfile) {
                      toggleDropdown();
                    }
                  }}
                />
              )}
              menuItems={(closeDropdown) => (
                <MenuItems
                  parentRef={avatarRef}
                  parentContainRef={avatarRef}
                  closeDropdown={closeDropdown}
                  style={{ paddingTop: '0px', paddingBottom: '0px' }}
                >
                  {
                    renderUserProfile
                      ? (
                        renderUserProfile({
                          user: sender,
                          close: closeDropdown,
                        })
                      )
                      : (
                        <UserProfile
                          user={sender}
                          onSuccess={closeDropdown}
                          disableMessaging
                        />
                      )
                  }
                </MenuItems>
              )}
            />
          )
        }
      </div>
      <div className="sendbird-openchannel-user-message__right">
        {
          !chainTop && (
            <div className="sendbird-openchannel-user-message__right__top">
              <Label
                className="sendbird-openchannel-user-message__right__top__sender-name"
                type={LabelTypography.CAPTION_2}
                color={isOperator ? LabelColors.SECONDARY_3 : LabelColors.ONBACKGROUND_2}
              >
                {
                  sender && (
                    sender.friendName
                    || sender.nickname
                    || sender.userId
                  )
                }
              </Label>
              <Label
                className="sendbird-openchannel-user-message__right__top__sent-at"
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_3}
              >
                {
                  message?.createdAt && (
                    format(message?.createdAt, 'p', {
                      locale: dateLocale,
                    })
                  )
                }
              </Label>
            </div>
          )
        }
        <div className="sendbird-openchannel-user-message__right__bottom">
          <Label
            className="sendbird-openchannel-user-message__right__bottom__message"
            type={LabelTypography.BODY_1}
            color={LabelColors.ONBACKGROUND_1}
          >
            {message?.message}
            {isEditedMessage(message) && (
              <Label
                key={uuidv4()}
                type={LabelTypography.BODY_1}
                color={LabelColors.ONBACKGROUND_2}
                calssName="sendbird-openchannel-user-message-word"
              >
                {` ${stringSet.MESSAGE_EDITED} `}
              </Label>
            )}
          </Label>
        </div>
        {
          (isPending || isFailed) && (
            <div className="sendbird-openchannel-user-message__right__tail">
              {
                isPending && (
                  <Loader
                    width="16px"
                    height="16px"
                  >
                    <Icon
                      className="sendbird-openchannel-user-message__right__tail__pending"
                      type={IconTypes.SPINNER}
                      fillColor={IconColors.PRIMARY}
                      width="16px"
                      height="16px"
                    />
                  </Loader>
                )
              }
              {
                isFailed && (
                  <Icon
                    className="sendbird-openchannel-user-message__right__tail__failed"
                    type={IconTypes.ERROR}
                    fillColor={IconColors.ERROR}
                    width="16px"
                    height="16px"
                  />
                )
              }
            </div>
          )
        }
      </div>
      {
        <div
          className="sendbird-openchannel-user-message__context-menu"
          ref={contextMenuRef}
          style={contextStyle}
        >
          <ContextMenu
            menuTrigger={(toggleDropdown) => (
              showMenuTrigger({ message: message, userId: userId, status: status }) && (
                <IconButton
                  className="sendbird-openchannel-user-message__context-menu--icon"
                  width="32px"
                  height="32px"
                  onClick={() => {
                    toggleDropdown();
                  }}
                >
                  <Icon
                    type={IconTypes.MORE}
                    fillColor={IconColors.CONTENT_INVERSE}
                    width="24px"
                    height="24px"
                  />
                </IconButton>
              )
            )}
            menuItems={(closeDropdown) => (
              <MenuItems
                parentRef={contextMenuRef}
                parentContainRef={contextMenuRef}
                closeDropdown={closeDropdown}
                openLeft
              >
                {
                  isFineCopy({ message: message, userId: userId, status: status }) && (
                    <MenuItem
                      className="sendbird-openchannel-user-message__context-menu__copy"
                      onClick={() => {
                        copyToClipboard(message.message);
                        closeDropdown();
                      }}
                    >
                      {stringSet.CONTEXT_MENU_DROPDOWN__COPY}
                    </MenuItem>
                  )
                }
                {
                  isFineEdit({ message: message, userId: userId, status: status }) && (
                    <MenuItem
                      className="sendbird-openchannel-user-message__context-menu__edit"
                      onClick={() => {
                        if (disabled) {
                          return;
                        }
                        showEdit(true);
                        closeDropdown();
                      }}
                    >
                      {stringSet.CONTEXT_MENU_DROPDOWN__EDIT}
                    </MenuItem>
                  )
                }
                {
                  isFineResend({ message: message, userId: userId, status: status }) && (
                    <MenuItem
                      className="sendbird-openchannel-user-message__context-menu__resend"
                      onClick={() => {
                        resendMessage(message);
                        closeDropdown();
                      }}
                    >
                      {stringSet.CONTEXT_MENU_DROPDOWN__RESEND}
                    </MenuItem>
                  )
                }
                {
                  isFineDelete({ message: message, userId: userId, status: status }) && (
                    <MenuItem
                      className="sendbird-openchannel-user-message__context-menu__delete"
                      onClick={() => {
                        if (disabled) {
                          return;
                        }
                        showRemove(true);
                        closeDropdown();
                      }}
                    >
                      {stringSet.CONTEXT_MENU_DROPDOWN__DELETE}
                    </MenuItem>
                  )
                }
              </MenuItems>
            )}
          />
        </div>
      }
    </div>
  );
}
