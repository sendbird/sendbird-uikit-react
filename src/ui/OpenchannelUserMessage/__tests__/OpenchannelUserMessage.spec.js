import React from 'react';
import { render, screen } from '@testing-library/react';

import OpenchannelUserMessage from "../index";

// mock date-fns to avoid problems from snapshot timestamping
// between testing in different locations
// ideally we want to mock date-fns globally - needs more research
jest.mock('date-fns/format', () => () => ('mock-date'));

const userId = 'hh-1234';
const getMockMessage = (callback) => {
  const mockMessage = {
    message: 'I am message',
    messageType: 'user',
    updatedAt: 0,
    sender: {
      profileUrl: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
      friendName: 'Hoon Baek',
      nickname: 'Hoon',
      userId,
    },
    createdAt: 1000,
    isResendable: () => { },
  };
  if (callback && typeof callback === 'function') {
    callback(mockMessage);
  }
  return mockMessage;
};

describe('ui/OpenchannelUserMessage', () => {
  it('should have default elements', function () {
    const { container } = render(
      <OpenchannelUserMessage
        message={getMockMessage()}
        userId={userId}
        status="succeeded"
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message')[0].className
    ).toContain('sendbird-openchannel-user-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__left__avatar').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__top__sender-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__top__sent-at').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__bottom__message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should not have chainTop elements', function () {
    const { container } = render(
      <OpenchannelUserMessage
        message={getMockMessage()}
        userId={userId}
        status="succeeded"
        chainTop
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message')[0].className
    ).toContain('sendbird-openchannel-user-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__left').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__left__avatar').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__top__sender-name').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__top__sent-at').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__bottom__message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should render pending icon if status is pending', function () {
    const { container } = render(
      <OpenchannelUserMessage
        message={getMockMessage((message) => {
          message.isResendable = () => true;
          message.sendingStatus = 'pending';
        })}
        userId={userId}
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__pending').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should render failed icon if status is failed', function () {
    const { container } = render(
      <OpenchannelUserMessage
        message={getMockMessage((message) => {
          message.isResendable = () => true;
          message.sendingStatus = 'failed';
        })}
        userId={userId}
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-user-message__right__tail__failed').length
    ).toBe(1);
  });

  it('should do a snapshot test of the OpenchannelUserMessage DOM', function () {
    const { asFragment } = render(
      <OpenchannelUserMessage
        message={{
          message: 'I am message',
          messageType: 'user',
          updatedAt: 0,
          sender: {
            profileUrl: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
            friendName: 'Hoon Baek',
            nickname: 'Hoon',
            userId: 'hh-1234',
          },
          createdAt: 1000,
          isResendable: () => { },
        }}
        status="succeeded"
        userId="hh-1234"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
