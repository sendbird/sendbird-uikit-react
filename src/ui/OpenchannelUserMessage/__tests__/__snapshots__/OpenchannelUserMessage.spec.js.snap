// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/OpenchannelUserMessage should do a snapshot test of the OpenchannelUserMessage DOM 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-openchannel-user-message"
  >
    <div
      class="sendbird-openchannel-user-message__left"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <div
          class="sendbird-openchannel-user-message__left__avatar sendbird-avatar"
          role="button"
          style="height: 28px; width: 28px; z-index: 0; left: 0px;"
          tabindex="0"
        >
          <div
            class="sendbird-avatar-img sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, 28px); max-width: 400px;"
          >
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, 28px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_12_512px.png);"
            />
            <img
              alt=""
              class="sendbird-image-renderer__hidden-image-loader"
              src="https://static.sendbird.com/sample/profiles/profile_12_512px.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-user-message__right"
    >
      <div
        class="sendbird-openchannel-user-message__right__top"
      >
        <span
          class="sendbird-openchannel-user-message__right__top__sender-name sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-2"
        >
          Hoon Baek
        </span>
        <span
          class="sendbird-openchannel-user-message__right__top__sent-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-3"
        >
          mock-date
        </span>
      </div>
      <div
        class="sendbird-openchannel-user-message__right__bottom"
      >
        <span
          class="sendbird-openchannel-user-message__right__bottom__message sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
        >
          I am message
        </span>
      </div>
    </div>
    <div
      class="sendbird-openchannel-user-message__context-menu"
      style="top: 2px;"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <button
          class="sendbird-openchannel-user-message__context-menu--icon sendbird-iconbutton "
          style="height: 32px; width: 32px;"
          type="button"
        >
          <span
            class="sendbird-iconbutton__inner"
          >
            <div
              class=" sendbird-icon sendbird-icon-more sendbird-icon-color--content-inverse"
              role="button"
              style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
              tabindex="0"
            >
              <test-file-stub />
            </div>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
